linalg/dual.hpp

```diff
@@ -0,0 +1,413 @@
+     double b)                                                  \
+  {                                                             \
+    return a.value x b;                                         \
+  }                                                             \
+                                                                \
+  template <typename value_type, typename gradient_type>        \
+  MFEM_HOST_DEVICE constexpr bool operator x(                   \
+     double a,                                                  \
+     const dual<value_type, gradient_type>& b)                  \
+  {                                                             \
+    return a x b.value;                                         \
+  }                                                             \
+                                                                \
+  template <typename value_type_a,                              \
+            typename gradient_type_a,                           \
+            typename value_type_b,                              \
+            typename gradient_type_b> MFEM_HOST_DEVICE          \
+  constexpr bool operator x(                                    \
+     const dual<value_type_a, gradient_type_a>& a,              \
+     const dual<value_type_b, gradient_type_b>& b)              \
+  {                                                             \
+    return a.value x b.value;                                   \
+  }
+
+binary_comparator_overload(<)   ///< implement operator<  for dual numbers
+binary_comparator_overload(<=)  ///< implement operator<= for dual numbers
+binary_comparator_overload(==)  ///< implement operator== for dual numbers
+binary_comparator_overload(>=)  ///< implement operator>= for dual numbers
+binary_comparator_overload(>)   ///< implement operator>  for dual numbers
+
+#undef binary_comparator_overload
+
+/** @brief compound assignment (+) for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type>& operator+=(dual<value_type, gradient_type>& a,
+                                            const dual<value_type, gradient_type>& b)
+{
+   a.value += b.value;
+   a.gradient += b.gradient;
+   return a;
+}
+
+/** @brief compound assignment (-) for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type>& operator-=(dual<value_type, gradient_type>& a,
+                                            const dual<value_type, gradient_type>& b)
+{
+   a.value -= b.value;
+   a.gradient -= b.gradient;
+   return a;
+}
+
+/** @brief compound assignment (+) for dual numbers with `double` righthand side */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type>& operator+=(dual<value_type, gradient_type>& a,
+                                            double b)
```

Comment:
It is better to add a separate template parameter for `value_type_b`, if the type of `b` has to be different from `value_type`.

Reply:
this suggestion would lead to `operator+=` being defined for any possible righthand side (strings, meshes, etc), which seems like a bad idea to me